api/types/saml.go

```diff
@@ -383,165 +369,48 @@ func (o *SAMLConnectorV2) GetTraitMappings() TraitMappingSet {
+	return o.Spec.SigningKeyPair
+}
+
+// SetSigningKeyPair sets signing key pair
+func (o *SAMLConnectorV2) SetSigningKeyPair(k *SigningKeyPair) {
+	o.Spec.SigningKeyPair = k
+}
+
+// CheckAndSetDefaults checks and sets default values
+func (o *SAMLConnectorV2) CheckAndSetDefaults() error {
+	if err := o.Metadata.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
 	}
 	if o.Metadata.Name == teleport.Local {
-		return nil, trace.BadParameter("ID: invalid connector name %v is a reserved name", teleport.Local)
+		return trace.BadParameter("ID: invalid connector name %v is a reserved name", teleport.Local)
 	}
 	if o.Spec.AssertionConsumerService == "" {
-		return nil, trace.BadParameter("missing acs - assertion consumer service parameter, set service URL that will receive POST requests from SAML")
+		return trace.BadParameter("missing acs - assertion consumer service parameter, set service URL that will receive POST requests from SAML")
 	}
 	if o.Spec.ServiceProviderIssuer == "" {
 		o.Spec.ServiceProviderIssuer = o.Spec.AssertionConsumerService
 	}
 	if o.Spec.Audience == "" {
 		o.Spec.Audience = o.Spec.AssertionConsumerService
 	}
-	certStore := dsig.MemoryX509CertificateStore{
-		Roots: []*x509.Certificate{},
-	}
-	// if we have a entity descriptor url, fetch it first
-	if o.Spec.EntityDescriptorURL != "" {
-		resp, err := http.Get(o.Spec.EntityDescriptorURL)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		if resp.StatusCode != http.StatusOK {
-			return nil, trace.BadParameter("status code %v when fetching from %q", resp.StatusCode, o.Spec.EntityDescriptorURL)
-		}
-		defer resp.Body.Close()
-		body, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			return nil, trace.Wrap(err)
-		}
-		o.Spec.EntityDescriptor = string(body)
-		log.Debugf("[SAML] Successfully fetched entity descriptor from %q", o.Spec.EntityDescriptorURL)
-	}
-	if o.Spec.EntityDescriptor != "" {
-		metadata := &types.EntityDescriptor{}
-		err := xml.Unmarshal([]byte(o.Spec.EntityDescriptor), metadata)
-		if err != nil {
-			return nil, trace.Wrap(err, "failed to parse entity_descriptor")
-		}
-
-		for _, kd := range metadata.IDPSSODescriptor.KeyDescriptors {
-			for _, samlCert := range kd.KeyInfo.X509Data.X509Certificates {
-				certData, err := base64.StdEncoding.DecodeString(strings.TrimSpace(samlCert.Data))
-				if err != nil {
-					return nil, trace.Wrap(err)
-				}
-				cert, err := x509.ParseCertificate(certData)
-				if err != nil {
-					return nil, trace.Wrap(err, "failed to parse certificate in metadata")
-				}
-				certStore.Roots = append(certStore.Roots, cert)
-			}
-		}
-		o.Spec.Issuer = metadata.EntityID
-		if len(metadata.IDPSSODescriptor.SingleSignOnServices) > 0 {
-			o.Spec.SSO = metadata.IDPSSODescriptor.SingleSignOnServices[0].Location
+	// Issuer and SSO can be automatically set later if EntityDescriptor is provided
+	if o.Spec.EntityDescriptorURL == "" && o.Spec.EntityDescriptor == "" {
```

Comment:
You should fail if at least one of these is not specified saying an entity descriptor (or entity descriptor URL) is required.

Reply:
This breaks the functionality of providing `Issuer` and or `SSO` but no `EntityDescriptor`, which is at the very least used by tests. Since this would break tests I'm going to leave it functionally as is for now. I did made this check a bit more compact, as it is re-validated by `ValidateSAMLConnector` after parsing the `EntityDescriptor`.